### PR TITLE
Add SMB1/SMB2

### PR DESCRIPTION
**Pull request by gpotter2**
- answers to SMB2 protocol requests
- answers SMB1 protocol requests
- adds "generic" packet classes: "Packet" and "PacketDissector". This could probably be abstracted a bit more, and generalized to other masscanned protos

### SMB1
```
smbclient -U "" -N -L 10.1.1.1 -d10 --option='client min protocol=NT1'
```
image
image

### SMB2
```
smbclient -U "" -N -L 10.1.1.1 -d10
```
image
image

### CODE REVIEW

File: src/proto/smb.rs [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+//
+// Masscanned is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+// or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+// License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Masscanned. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashSet;
```
Comment (Frky) [inline src/proto/smb.rs:17]: We *try* to make import orders consistant in this projects (although there might be some mistakes in some files) ; could you please try to do the same here?
(To check, but irrc the order should be:
- log + std imports
- other lib imports
- local project imports)

File: src/proto/smb.rs [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+pub const SMB2_PATTERN_MAGIC: &[u8; 8] = b"\x00\x00**\xfeSMB";
+
+// Build/Dissect secblob with Scapy using: GSSAPI_BLOB(b"`\x82.....")
+const SECURITY_BLOB: &[u8; 320] = b"`\x82\x01<\x06\x06+\x06\x01\x05\x05\x02\xa0\x82\x0100\x82\x01,\xa0\x1a0\x18\x06\n+\x06\x01\x04\x01\x827\x02\x02\x1e\x06\n+\x06\x01\x04\x01\x827\x02\x02\n\xa2\x82\x01\x0c\x04\x82\x01\x08NEGOEXTS\x01\x00\x00\x00\x00\x00\x00\x00`\x00\x00\x00p\x00\x00\x001<*:\xc7+<\xa9m\xac8t\xa7\xdd\x1d[\xf4Rk\x17\x03\x8aK\x91\xc2\t}\x9a\x8f\xe6,\x96\\Q$/\x90MG\xc7\xad\x8f\x87k\"\x02\xbf\xc6\x00\x00\x00\x00\x00\x00\x00\x00`\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\\3S\r\xea\xf9\rM\xb2\xecJ\xe3xn\xc3\x08NEGOEXTS\x03\x00\x00\x00\x01\x00\x00\x00@\x00\x00\x00\x98\x00\x00\x001<*:\xc7+<\xa9m\xac8t\xa7\xdd\x1d[\\3S\r\xea\xf9\rM\xb2\xecJ\xe3xn\xc3\x08@\x00\x00\x00X\x00\x00\x000V\xa0T0R0'\x80%0#1!0\x1f\x06\x03U\x04\x03\x13\x18Token Signing Public Key0'\x80%0#1!0\x1f\x06\x03U\x04\x03\x13\x18Token Signing Public Key";
+
+/////////////
+// Netbios //
+/////////////
+
+#[allow(dead_code)]
```
Comment (Frky) [inline src/proto/smb.rs:40]: Why is that needed?

File: src/proto/smb.rs [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+        resp.extend_from_slice(&((size & 0xffff) as u16).to_be_bytes());
+        Some(resp)
+    }
+}
+
+//////////
+// SMB1 //
+//////////
+
+#[allow(dead_code)]
```
Comment (Frky) [inline src/proto/smb.rs:81]: Why is that needed?

File: src/proto/smb.rs [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+        resp.extend_from_slice(&[0; 2]); // Reserved
+        resp.extend_from_slice(&self.tid.to_le_bytes()); // TID
+        resp.extend_from_slice(&self.pid_low.to_le_bytes()); // PIDLOW
+        resp.extend_from_slice(&self.uid.to_le_bytes()); // UID
+        resp.extend_from_slice(&self.mid.to_le_bytes()); // MID
+        Some(resp)
+    }
+}
+
+#[allow(dead_code)]
```
Comment (Frky) [inline src/proto/smb.rs:167]: Why is that needed?

**Review by Frky — CHANGES_REQUESTED** [review]
Thank you for the PR, and for implementing new protocols in **masscanned**. **SMB** is a very nice proto to add, as it is one of the most scanned application protocols being scanned from what we see, so this **PR** definitely implements something very useful.

However, I(d like to suggest some changes that are really important given how **masscanned** aims to work. As the protocol parsing is network-layers agnostic, one could assume that the payload can either come from `UDP` or `TCP` packets (or `SCTP`, etc.). This is really important to keep in mind, because it means that, over `TCP` for instance, **we should not assume** that the payload is coming all at once. The `SMB` payload could come byte by byte in different `TCP` packets ; and the protocol parsers must take that into consideration. That is the reason why `HTTP` parser, for instance, is implemented using a FSM.

To give you insights on how to improve your protocol parsers:

* you should only read one byte at the time, never the whole payload at once;
* to be sure of that, it is good to add a unit test for your parser where you call `repl` giving the payload byte by byte (and it should only respond with new data when the payload has been completely been parsed);
* you can store *states* in a `ProtocolState` structure (or something that derives from that) - see PR #21 (WIP).

File: src/proto/smb.rs [hunk 5]
```diff
@@ -0,0 +1,1150 @@
...
+//////////////
+// Handlers //
+//////////////
+
+pub fn repl_smb1<'a>(
+    data: &'a [u8],
+    _masscanned: &Masscanned,
+    _client_info: &ClientInfo,
+) -> Option<Vec<u8>> {
+    let mut nbtsession: NBTSession<SMB1Header> = NBTSession::new();
```
Comment (gpotter2) [inline src/proto/smb.rs:963]: Note: this is re-created on each call to `repl_smb1` which is not expected: we should store it somewhere in `client_info`. Waiting for PR #21 as discussed with frk

File: src/proto/smb.rs [hunk 6]
```diff
@@ -0,0 +1,1150 @@
...
+    }
+    return nbtsession.repl();
+}
+
+pub fn repl_smb2<'a>(
+    data: &'a [u8],
+    _masscanned: &Masscanned,
+    _client_info: &ClientInfo,
+) -> Option<Vec<u8>> {
+    let mut nbtsession: NBTSession<SMB2Header> = NBTSession::new();
```
Comment (gpotter2) [inline src/proto/smb.rs:975]: Same

File: src/proto/smb.rs [hunk 7]
```diff
@@ -0,0 +1,1150 @@
...
+            84, 111, 107, 101, 110, 32, 83, 105, 103, 110, 105, 110, 103, 32, 80, 117, 98, 108,
+            105, 99, 32, 75, 101, 121,
+        ];
+        assert_eq!(answer[..0x3c], expected[..0x3c]); // Test equality except "ServerTime" field
+        assert_eq!(answer[0x3c + 8..], expected[0x3c + 8..]);
+    }
+    #[test]
+    fn test_smb2_protocol_nego_parsing() {
+        let mut nbtsession: NBTSession<SMB2Header> = NBTSession::new();
+        nbtsession.parse_all(SMB2_REQ_PAYLOAD);
```
Comment (gpotter2) [inline src/proto/smb.rs:1082]: But in tests, `parse_all` calls `parse()` byte per byte: dissection works.

**Review by Frky — APPROVED** [review]

File: src/proto/smb.rs [hunk 8]
```diff
@@ -0,0 +1,1143 @@
...
+
+// Build/Dissect secblob with Scapy using: GSSAPI_BLOB(b"`\x82.....")
+const SECURITY_BLOB: &[u8; 320] = b"`\x82\x01<\x06\x06+\x06\x01\x05\x05\x02\xa0\x82\x0100\x82\x01,\xa0\x1a0\x18\x06\n+\x06\x01\x04\x01\x827\x02\x02\x1e\x06\n+\x06\x01\x04\x01\x827\x02\x02\n\xa2\x82\x01\x0c\x04\x82\x01\x08NEGOEXTS\x01\x00\x00\x00\x00\x00\x00\x00`\x00\x00\x00p\x00\x00\x001<*:\xc7+<\xa9m\xac8t\xa7\xdd\x1d[\xf4Rk\x17\x03\x8aK\x91\xc2\t}\x9a\x8f\xe6,\x96\\Q$/\x90MG\xc7\xad\x8f\x87k\"\x02\xbf\xc6\x00\x00\x00\x00\x00\x00\x00\x00`\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\\3S\r\xea\xf9\rM\xb2\xecJ\xe3xn\xc3\x08NEGOEXTS\x03\x00\x00\x00\x01\x00\x00\x00@\x00\x00\x00\x98\x00\x00\x001<*:\xc7+<\xa9m\xac8t\xa7\xdd\x1d[\\3S\r\xea\xf9\rM\xb2\xecJ\xe3xn\xc3\x08@\x00\x00\x00X\x00\x00\x000V\xa0T0R0'\x80%0#1!0\x1f\x06\x03U\x04\x03\x13\x18Token Signing Public Key0'\x80%0#1!0\x1f\x06\x03U\x04\x03\x13\x18Token Signing Public Key";
+
+////////////
+// Common //
+////////////
+
+/// ### PacketDissector
+/// A util class used to dissecate fields.
```
Comment (Frky) [inline src/proto/smb.rs:40]: s/dissecate/dissect/

File: src/proto/smb.rs [hunk 9]
```diff
@@ -0,0 +1,1143 @@
...
+pub fn repl_smb1<'a>(
+    data: &'a [u8],
+    _masscanned: &Masscanned,
+    _client_info: &ClientInfo,
+) -> Option<Vec<u8>> {
+    let mut nbtsession: NBTSession<SMB1Header> = NBTSession::new();
+    for byte in data {
+        nbtsession.parse(byte);
+    }
+    return nbtsession.repl();
```
Comment (Frky) [inline src/proto/smb.rs:959]: More rusty to use `nbtsession.repl()` (no `return`, no `;`) at the end of the function I think, do you agree?

File: src/proto/smb.rs [hunk 10]
```diff
@@ -0,0 +1,1143 @@
...
+pub fn repl_smb2<'a>(
+    data: &'a [u8],
+    _masscanned: &Masscanned,
+    _client_info: &ClientInfo,
+) -> Option<Vec<u8>> {
+    let mut nbtsession: NBTSession<SMB2Header> = NBTSession::new();
+    for byte in data {
+        nbtsession.parse(byte);
+    }
+    return nbtsession.repl();
```
Comment (Frky) [inline src/proto/smb.rs:971]: Same remark here.

File: src/proto/smb.rs [hunk 11]
```diff
@@ -0,0 +1,1143 @@
...
+    fn next_state(&mut self, state: T) {
+        self.state = state;
+        self.i = 0;
+    }
+    fn next_state_when_i_reaches(&mut self, state: T, i: usize) {
+        if self.i == i {
+            self.next_state(state);
+        }
+    }
+    fn _read_u(&mut self, byte: &u8, value: usize, next_state: T, size: usize) -> usize {
```
Comment (Frky) [inline src/proto/smb.rs:62]: Why not naming this function `_read_usize`?

File: src/proto/smb.rs [hunk 12]
```diff
@@ -0,0 +1,1143 @@
...
+    }
+    fn read_ule32(&mut self, byte: &u8, value: u32, next_state: T) -> u32 {
+        self._read_ule(byte, value as usize, next_state, 4) as u32
+    }
+    fn read_ule64(&mut self, byte: &u8, value: u64, next_state: T) -> u64 {
+        self._read_ule(byte, value as usize, next_state, 8) as u64
+    }
+}
+
+pub trait Packet {
```
Comment (Frky) [inline src/proto/smb.rs:87]: This name seems to be unfortunate, as there already is a Packet trait in the library `pnet` that this project uses lots (https://docs.rs/pnet/latest/pnet/packet/trait.Packet.html).

Comment (gpotter2) [reply src/proto/smb.rs:87]: Do you think that's an issue? Rust handles quite well naming conflicts.

File: src/proto/smb.rs [hunk 13]
```diff
@@ -0,0 +1,1143 @@
...
+            SMB1NegotiateRequestState::End => {}
+        }
+    }
+
+    fn repl(&self) -> Option<Vec<u8>> {
+        if !matches!(self.d.state, SMB1NegotiateRequestState::End) {
+            return None;
+        }
+        let mut resp: Vec<u8> = Vec::new();
+        let time: u64 = (EPOCH_1601 + chrono::Utc::now().timestamp() as u64) * (1e7 as u64);
```
Comment (Frky) [inline src/proto/smb.rs:414]: You can use `SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)` (see https://doc.rust-lang.org/std/time/struct.SystemTime.html).

**Review by Frky — COMMENTED** [review]
You need to fix the tests (at least cargo fmt is failing).

File: src/proto/smb.rs (hunk 12 above)
Comment (Frky) [reply src/proto/smb.rs:87]: Yes, but if this trait is to be used in other protocol dissectors, I think it could be very confusing for humans, don't you think?

File: src/proto/smb.rs (hunk 13 above)
Comment (gpotter2) [reply src/proto/smb.rs]: Will do.
I had found chrono in Cargo.toml so apparently this is used somewhere :P

I looked into using something like `duration_since(...1601 EPOCH...)` but the format [looks kinda crazy](https://github.com/rust-lang/rust/blob/9a60099cc43c8a07abb280be323d1ed9afc27f2c/library/std/src/sys/windows/time.rs#L27)

File: src/proto/smb.rs (hunk 12 above)
Comment (gpotter2) [reply src/proto/smb.rs]: Fair enough, I've renamed it MPacket for "Masscanned-packet"

**Comment by gpotter2** [comment]
- formatting issues addressed
- rebased on top of master :P

**Comment by gpotter2** [comment]
I've added the functional test to https://github.com/ivre/masscanned/pull/43 because I ended up splitting the test suite :P

**Comment by Frky** [comment]
Thank you for the PR.
